adoptium-updater-parent/adoptium-datasources-parent/adoptium-github-datasource/src/main/kotlin/net/adoptium/api/v3/dataSources/github/GitHubAuth.kt

```diff
@@ -1,22 +1,57 @@
 package net.adoptium.api.v3.dataSources.github
 
-import org.slf4j.LoggerFactory
 import java.io.File
 import java.nio.file.Files
 import java.util.Properties
+import org.slf4j.LoggerFactory
+import io.jsonwebtoken.Jwts
+import io.jsonwebtoken.SignatureAlgorithm
+import org.kohsuke.github.GHAppInstallation
+import org.kohsuke.github.GHAppInstallationToken
+import java.security.KeyFactory
+import java.security.spec.PKCS8EncodedKeySpec
+import java.util.Base64
+import java.util.Date
+import org.kohsuke.github.GitHub
+import org.kohsuke.github.GitHubBuilder
 
 class GitHubAuth {
+    data class AuthInfo(val token: String?, val type: AuthType, val expirationTime: Date?)
+    enum class AuthType {
+        APP, TOKEN, NONE
+    }
 
     companion object {
         @JvmStatic
         private val LOGGER = LoggerFactory.getLogger(this::class.java)
 
-        fun readToken(): String? {
-            var token = System.getenv("GITHUB_TOKEN")
-            if (token.isNullOrEmpty()) {
-                token = System.getProperty("GITHUB_TOKEN")
+        fun getAuthenticationToken(): AuthInfo {
+            val appId = System.getenv("GITHUB_APP_ID")
+            val privateKey = System.getenv("GITHUB_APP_PRIVATE_KEY")
+            val installationId = System.getenv("GITHUB_APP_INSTALLATION_ID")
+
+            return if (!appId.isNullOrEmpty() && !privateKey.isNullOrEmpty() && !installationId.isNullOrEmpty()) {
+                LOGGER.info("Using GitHub App for authentication")
+                val token = authenticateAsGitHubApp(appId, privateKey, installationId)
+                if (token != null) {
+                    AuthInfo(token.token, AuthType.APP, token.expiresAt)
+                } else {
+                    LOGGER.error("Failed to authenticate as GitHub App")
+                    AuthInfo(null, AuthType.NONE, null)
```

Comment:
not sure if there is any point continuing executing at this point? Throw an error and remove NONE type?

Reply:
In the past we've always supported running the API without any tokens. I wanted to preserve that behavior?